keps/sig-architecture/20190412-conformance-behaviors.md

```diff
@@ -0,0 +1,268 @@
+  - id: lifecycle/submit-remove
+    description: When a Pod is be created with a unique label, it can be queried using the label selector. It can be watched and will more to Running state. The Pod can be deleted, which sets the pod deletion timestamp, and the watch will receive a deletion event. Afterward, the query with the label selector must be empty.
+  - id: pods/lifecycle/label-update
+    description: Updating a Pod label is successful.
+  - id: pods/lifecycle/active-deadline
+    description: The activeDeadlineSeconds is honored when updated on the podSpec.
+- category: Graceful Termination
+  behaviors:
+  - id: pods/graceful/podspec-period-honored
+    description: The terminationGracePeriodSeconds as specified in the podSpec is honored.
+  - id: pods/graceful/request-period-honored
+    description: The terminationGracePeriod as specified in the DELETE request is honored.
+  - id: pods/graceful/with-prestop
+    description: The terminationGracePeriod is honored when a preStop hook blocks.
+- category: Images
+  behaviors:
+  - id: pods/images/update
+    description: When a podSpec container image is updated, the image is pulled and the container restarted.
+  - id: pods/images/pull-policy
+    description: The podSpec imagePullPolicy is honored during creation and update.
+- category: OS Namespaces
+  behaviors:
+  - id: pods/os-namespaces/network-default
+    description: By default, containers in a pod share the same network namespace.
+  - id: pods/os-namespaces/network-host
+    description: Containers share the host network namespace when hostNetwork is true.
+  - id: pods/os-namespaces/ipc-default
+    description: By default, containers in a pod share the same IPC namespace.
+  - id: pods/os-namespaces/ipc-host
+    description: Containers share the host IPC namespace when hostIPC is true.
+  - id: pods/os-namespaces/pid-default
+    description: By default, containers are in their own PID namespace.
+  - id: pods/os-namespaces/pid-host
+    description: Containers share the host PID namespace when hostPID is true.
+- category: Volumes
+  behaviors:
+  - id: pods/volumes/sa-token
+    description: Pods by default have a service account token mounted as a volume.
+  - id: pods/volumes/sa-token-disabled
+    description: No service account token volume is mounted if automountServiceAccountToken is false.
+  - id: pods/volumes/shared
+    description: Two containers can mount the same volume and see each others' files.
+```
+
+In order to allow incremental adoption of this framework, rather than a new
+conformance framework function, test authors will indicate the behaviors covered
+by their tests with a label. So, a change may look something like:
+
+```
+-       framework.ConformanceIt("should get a host IP [NodeConformance]", func() {
++       framework.ConformanceIt("should get a host IP [NodeConformance] [Behavior:pods/lifecycle/hostip]", func() {
```

Comment:
At the end of the day you're basically outlining a better description...  Should we just update the current ones to spell out the behavior using a clear defined pattern?

Reply:
One goal is to separate the people reviewing the behaviors from those that review the tests. I think we could build out the entire list of behaviors for, say, the Pod API pretty quickly. Then, we would backfill with tests / annotate existing tests with these IDs. We get a measure of coverage of the behaviors, plus a clear consensus on what "conforming behaviors" are.

Just updating current tests means figuring out what each test does, then documenting it. Which doesn't leave us with any notion of how much of the real behavior is covered.